### ISSUE TITLE
Srcsets, targetwidths, and server-side rendering

### ISSUE THREAD

**Issue by tstirrat15** [imgix/react-imgix#285]
- [x] Please search through the existing issues (both open AND closed) to see if your question has already been discussed. Github issue search can be used for this: https://github.com/imgix/react-imgix/issues?utf8=%E2%9C%93&q=is%3Aissue

**Question**
I saw  in [another issue](https://github.com/imgix/react-imgix/issues/71#issuecomment-276523785) that one of the goals of react-imgix is to generate a srcset that corresponds pretty well to the rendered size of the image. This works well if the code is first being executed on the frontend and is already in a browser context, but it seems that the backend behavior is to grab [every size referenced in targetWidths.js](https://github.com/imgix/react-imgix/blob/master/src/targetWidths.js) and generate a srcset for every width. In our experience, that makes for a whole lot of markup, with sizes ranging from hundreds of pixels to thousands of pixels that aren't always relevant.

* Is there a reason for the particular granularity specified in `targetWidths`, or could those be modified?
* Are there ways to give hints to the server render about the user's viewport in a way that would produce a smaller srcset?
* For sizes, we mostly fill the width of the parent container and use `100vw` as our size prop. Is there something that we could do differently there that would produce smaller srcsets?

**Comment by jayeb** [comment]
Hi Tanner, thanks for the ticket.

The issue you're referencing is two years old and that answer, while it was accurate back then (5.2.0, I believe), no longer reflects the behavior of the library.

In 8.0.0 we moved from a "measure the DOM and download the best-sized image to fit" approach to a "generate a robust `srcset` attribute and let the browser do the work" approach. This has a number of benefits:

- It works just as well server-side as it does on the browser-side.
- It's more performant for the client, since it doesn't rely on timers or events to select images.
- It's more performant in terms of download speed, since it better leverage's imgix's cache of rendered images.

Thus, the behavior you're seeing on the backend is now also the behavior on the front-end: we simply use imgix's parameters to generate a massive "menu" of sizes for the browser to choose from and list them all out in the `srcset` attribute. The first version of this behavior was somewhat naive, in retrospect--we generated our list of sizes by combining a (somewhat outdated) list of common device and screen sizes with a stepped list of dimensions from 100 to 4000.

In 8.4.0, however, we refined this approach, and now generate a list of resolutions that starts at 100 and scales up by 16% every step, all the way up to 8192 (the maximum image size imgix can render). This keeps the srcset to a much more reasonable size (30 sizes), while maintaining rendered image quality by ensuring that a client will never download an image that's more than 8% larger or smaller than its rendered size.

This approach is not currently tunable, but we have every intention of adding configuration parameters that would allow you to override the minimum and maximum sizes (100 and 8192 currently), and tune the 16% step rate. We're focusing our SDK efforts elsewhere at the moment, but we will be returning to this eventually. In the meantime, if you're like to take a shot at adding these configuration options yourself we'd love to review pull requests.

Hope this helps! I'm gonna close this ticket, but if you have further questions or ideas I'd be happy to continue the conversation.
